### PR TITLE
Backfill query implementation

### PR DESCRIPTION
**Pull request by yeukovichd**
**What this PR does / Why we need it**:
PR allows querying and caching backfills.

### CODE REVIEW

**Comment by aLekSer** [comment]
This might be helpful to see the diff, as new file was added `cache.go` and it is hard to see that no previous functionality was changed:
https://gist.github.com/aLekSer/392cf3b4cf8cb2c754b632a635c3e7f4

File: internal/app/query/cache.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
+package query
```
Comment (aLekSer) [inline internal/app/query/cache.go:1]: License header is missing, please copy from another file.

File: internal/statestore/backfill.go [hunk 2]
```diff
@@ -98,6 +98,46 @@ func (rb *redisBackend) GetBackfill(ctx context.Context, id string) (*pb.Backfil
...
+		return nil, status.Errorf(codes.Unavailable, "GetBackfills, failed to connect to redis: %v", err)
+	}
+	defer handleConnectionClose(&redisConn)
+
+	queryParams := make([]interface{}, len(ids))
+	for i, id := range ids {
+		queryParams[i] = id
+	}
+
+	slices, err := redis.ByteSlices(redisConn.Do("MGET", queryParams...))
```
Comment (aLekSer) [inline internal/statestore/backfill.go:118]: Nice to see one call to get Multiple Backfills. Would be useful.

File: internal/app/query/cache.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+// gives a safe view into that map cache.
+type cache struct {
+	store    statestore.Service
+	requests chan *cacheRequest
+	// Single item buffered channel.  Holds a value when runQuery can be safely
+	// started.  Basically a channel/select friendly mutex around runQuery
+	// running.
+	startRunRequest chan struct{}
+	wg              sync.WaitGroup
+	// Mutlithreaded unsafe fields, only to be written by update, and read when
```
Comment (aLekSer) [inline internal/app/query/cache.go:27]:
```suggestion
	// Multithreaded unsafe fields, only to be written by update, and read when
```

File: internal/app/query/cache.go (hunk 1 above)
Comment (aLekSer) [reply internal/app/query/cache.go:1]: Might be more safe for the reviewers and version control to move this code back where it resided.

File: internal/app/query/cache.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	c.startRunRequest <- struct{}{}
+	b.AddHealthCheckFunc(c.store.HealthCheck)
+
+	return c
+}
+
+func updateTicketCache(store statestore.Service, value interface{}) error {
+	if value == nil {
+		return fmt.Errorf("expecting not nil value")
```
Comment (akremsa) [inline internal/app/query/cache.go:125]: I think you should return a proper error code here and in other places where you return an error

**Review by aLekSer — APPROVED** [review]
Run the test and extended it a bit, works great. Thanks for the PR. Left minor comments.

File: internal/statestore/backfill.go [hunk 5]
```diff
@@ -98,6 +98,46 @@ func (rb *redisBackend) GetBackfill(ctx context.Context, id string) (*pb.Backfil
...
+	if err != nil {
+		err = errors.Wrapf(err, "failed to lookup backfills %v", ids)
+		return nil, status.Errorf(codes.Internal, "%v", err)
+	}
+
+	list := make([]*pb.Backfill, 0, len(ids))
+	for i, s := range slices {
+		if s != nil {
```
Comment (akremsa) [inline internal/statestore/backfill.go:126]: should we log a case where the nil value is returned calling for a proper ID?

File: internal/testing/e2e/query_backfills_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		require.Nil(t, resp)
+		require.Equal(t, io.EOF, err)
+	}
+	{
+		stream, err := om.Query().QueryBackfills(context.Background(), &pb.QueryBackfillsRequest{Pool: pool})
+		require.NoError(t, err)
+		resp, err := stream.Recv()
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+		require.Equal(t, int64(firstBackfillGeneration), resp.Backfills[0].Generation)
```
Comment (akremsa) [inline internal/testing/e2e/query_backfills_test.go:143]: I think it's worth checking require.NotEmpty(t, resp.Backfills)

File: internal/testing/e2e/query_backfills_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		require.NoError(t, err)
+		require.Equal(t, int64(firstBackfillGeneration), resp.Generation)
+	}
+	{
+		stream, err := om.Query().QueryBackfills(context.Background(), &pb.QueryBackfillsRequest{Pool: pool})
+		require.NoError(t, err)
+		resp, err := stream.Recv()
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+		require.Equal(t, int64(firstBackfillGeneration), resp.Backfills[0].Generation)
```
Comment (akremsa) [inline internal/testing/e2e/query_backfills_test.go:175]: require.NotEmpty(t, resp.Backfills)

File: internal/testing/e2e/query_backfills_test.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+		require.NotNil(t, resp)
+		require.Equal(t, int64(firstBackfillGeneration+1), resp.Generation)
+	}
+	{
+		stream, err := om.Query().QueryBackfills(context.Background(), &pb.QueryBackfillsRequest{Pool: pool})
+		require.NoError(t, err)
+		resp, err := stream.Recv()
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+		require.Equal(t, int64(firstBackfillGeneration+1), resp.Backfills[0].Generation)
```
Comment (akremsa) [inline internal/testing/e2e/query_backfills_test.go:194]: require.NotEmpty(t, resp.Backfills)

File: internal/testing/e2e/query_backfills_test.go (hunk 6 above)
Comment (aLekSer) [reply internal/testing/e2e/query_backfills_test.go:143]: Or `require.Len(t, resp.Backfills, 1)` before this line.

**Comment by aLekSer** [comment]
Hello @Laremere, can you please review this change?

**Review by akremsa — APPROVED** [review]

File: internal/statestore/backfill.go [hunk 9]
```diff
@@ -299,19 +299,38 @@ func (rb *redisBackend) GetIndexedBackfills(ctx context.Context) (map[string]int
 	}
 	defer handleConnectionClose(&redisConn)
 
-	bfIndex, err := redis.StringMap(redisConn.Do("HGETALL", allBackfills))
+	ttl := getBackfillReleaseTimeout(rb.cfg)
+	curTime := time.Now()
+	endTimeInt := curTime.Add(time.Hour).UnixNano()
+	startTimeInt := curTime.Add(-ttl).UnixNano()
+
+	acknowledgedIds, err := redis.Strings(redisConn.Do("ZRANGEBYSCORE", backfillLastAckTime, startTimeInt, endTimeInt))
```
Comment (aLekSer) [inline internal/statestore/backfill.go:307]: Dmitry, can you please explain this, some comments.

**Comment by aLekSer** [comment]
@Laremere we have one of your commit which is merge, so CLA check is broken.

**Pull request closed without merging**